### PR TITLE
add/ignore typing in napari.utils.tree.group

### PR DESCRIPTION
**Pull request by Carreau**
Fairly straitforward. 

A lot of ignore as multiple inheritance make typing really tough in this case.

### CODE REVIEW

**Review by brisvag — APPROVED** [review]

**Comment by brisvag** [comment]
Due to the lots of ignore, would be nice to get some input from other typing experts (@dstansby ?) cause this is tricky for me :)

**Comment by Carreau** [comment]
No worries, I hope that once the full codebase is types we can review the ignore and update some of them.

File: napari/utils/tree/group.py [hunk 1]
```diff
@@ -70,17 +78,17 @@ def __newlike__(self, iterable: Iterable):
         new._list.extend(iterable)
         return new
 
-    def __getitem__(self, key) -> Union[NodeType, Group[NodeType]]:
+    def __getitem__(self, key) -> Union[NodeType, Group[NodeType]]:  # type: ignore [override]
         return super().__getitem__(key)
```
Comment (dstansby) [inline napari/utils/tree/group.py:84]:
```suggestion
```
Given this just calls `super().__getitem__`, I think it's safe to get rid of it and rely on typing from the parent class?

**Review by dstansby — COMMENTED** [review]
I had a look at the rest of the typing ignores comments added here, and agree that they're quite complex to work out! I left one suggestion though

File: napari/utils/tree/group.py (hunk 1 above)
Comment (Czaki) [reply napari/utils/tree/group.py:84]: but does the super class type annotation resolve to proper type?

File: napari/utils/tree/group.py [hunk 2]
```diff
@@ -44,7 +52,7 @@ def __init__(
     ) -> None:
         Node.__init__(self, name=name)
         SelectableNestableEventedList.__init__(
-            self,
+            self,  # type: ignore [arg-type]
```
Comment (Czaki) [inline napari/utils/tree/group.py:57]: what is error message here?

**Review by Czaki — CHANGES_REQUESTED** [review]
I think that this errors should be fixed, not ignored. It looks like some of them will happen in any `SelectableNestableEventedList`. So I do not agree on this ignoring decision and think that type annotation needt to be fixed in base classes. 

Here outputfrom mypy run without ignore:

```
napari/utils/tree/group.py:47: error: Argument 1 to "__init__" of "SelectableEventedList" has incompatible type "Group[NodeType]"; expected "SelectableEventedList[_T]"  [arg-type]
napari/utils/tree/group.py:73: error: Signature of "__getitem__" incompatible with supertype "NestableEventedList"  [override]
napari/utils/tree/group.py:73: note:      Superclass:
napari/utils/tree/group.py:73: note:          @overload
napari/utils/tree/group.py:73: note:          def __getitem__(self, int, /) -> NodeType | NestableEventedList[NodeType]
napari/utils/tree/group.py:73: note:          @overload
napari/utils/tree/group.py:73: note:          def __getitem__(self, ParentIndex, /) -> NestableEventedList[NodeType]
napari/utils/tree/group.py:73: note:          @overload
napari/utils/tree/group.py:73: note:          def __getitem__(self, slice, /) -> NestableEventedList[NodeType]
napari/utils/tree/group.py:73: note:          @overload
napari/utils/tree/group.py:73: note:          def __getitem__(self, tuple[int | slice, ...], /) -> NodeType | NestableEventedList[NodeType]
napari/utils/tree/group.py:73: note:      Subclass:
napari/utils/tree/group.py:73: note:          def __getitem__(self, Any, /) -> NodeType | Group[NodeType]
napari/utils/tree/group.py:73: error: Signature of "__getitem__" incompatible with supertype "TypedMutableSequence"  [override]
napari/utils/tree/group.py:73: note:      Superclass:
napari/utils/tree/group.py:73: note:          @overload
napari/utils/tree/group.py:73: note:          def __getitem__(self, int, /) -> NodeType
napari/utils/tree/group.py:73: note:          @overload
napari/utils/tree/group.py:73: note:          def __getitem__(self, slice, /) -> TypedMutableSequence[NodeType]
napari/utils/tree/group.py:73: note:      Subclass:
napari/utils/tree/group.py:73: note:          def __getitem__(self, Any, /) -> NodeType | Group[NodeType]
napari/utils/tree/group.py:73: error: Signature of "__getitem__" incompatible with supertype "MutableSequence"  [override]
napari/utils/tree/group.py:73: note:      Superclass:
napari/utils/tree/group.py:73: note:          @overload
napari/utils/tree/group.py:73: note:          def __getitem__(self, int, /) -> NodeType
napari/utils/tree/group.py:73: note:          @overload
napari/utils/tree/group.py:73: note:          def __getitem__(self, slice, /) -> MutableSequence[NodeType]
napari/utils/tree/group.py:73: note:      Subclass:
napari/utils/tree/group.py:73: note:          def __getitem__(self, Any, /) -> NodeType | Group[NodeType]
napari/utils/tree/group.py:73: error: Signature of "__getitem__" incompatible with supertype "Sequence"  [override]
napari/utils/tree/group.py:73: note:      Superclass:
napari/utils/tree/group.py:73: note:          @overload
napari/utils/tree/group.py:73: note:          def __getitem__(self, int, /) -> NodeType
napari/utils/tree/group.py:73: note:          @overload
napari/utils/tree/group.py:73: note:          def __getitem__(self, slice, /) -> Sequence[NodeType]
napari/utils/tree/group.py:73: note:      Subclass:
napari/utils/tree/group.py:73: note:          def __getitem__(self, Any, /) -> NodeType | Group[NodeType]
napari/utils/tree/group.py:81: error: Item "Node" of "NodeType | Group[NodeType]" has no attribute "__iter__" (not iterable)  [union-attr]
napari/utils/tree/group.py:83: error: Argument 1 to "__delitem__" of "EventedList" has incompatible type "int | slice | tuple[int | slice, ...]"; expected "int | slice"  [arg-type]
napari/utils/tree/group.py:104: error: Incompatible types in "yield" (actual type "Node", expected type "NodeType")  [misc]
napari/utils/tree/group.py:105: error: "Node" has no attribute "__iter__"; maybe "__str__"? (not iterable)  [attr-defined]
```

File: napari/utils/tree/group.py (hunk 2 above)
Comment (Carreau) [reply napari/utils/tree/group.py:57]:
```
Argument 1 to "__init__" of "SelectableEventedList" has incompatible type "Group[NodeType]"; expected "SelectableEventedList[_T]
```

File: napari/utils/tree/group.py (hunk 1 above)
Comment (Carreau) [reply napari/utils/tree/group.py:84]: No, it missed `Group[NodeType]` in the Union

Technically mypy does not complain so we could remove it until we actually get an error.

File: napari/utils/tree/group.py (hunk 2 above)
Comment (Carreau) [reply napari/utils/tree/group.py:57]: I pushed a commit with a comment

**Comment by Carreau** [comment]
> I think that this errors should be fixed, not ignored. It looks like some of them will happen in any `SelectableNestableEventedList`

I don't think they can be fixed, I think this is due to the fact that `Group` is multiple inheritance and that mypy can't infer that `Group` is a `NodeType`.

Even using `from typing_extensions import Self` and using `-> Self` in `SelectableEventedList` as return Type I guet crazy incompatible signatures:

```
 Signature of "__getitem__" incompatible with supertype "NestableEventedList"  [override]
      Superclass:
          @overload
          def __getitem__(self, int, /) -> NodeType | Group[NodeType]
          ...
      Subclass:
          def __getitem__(self, int, /) -> NodeType | Group[NodeType]
```

If you know how to fix those, please go ahead.

**Comment by Carreau** [comment]
Moving back to draft as I'm unsure how to fix.

**Pull request closed without merging**